reconcile/utils/ocm/clusters.py

```diff
@@ -0,0 +1,196 @@
+            subscription_filter=subscription_filter,
+        ).values()
+    )
+
+
+def discover_clusters_for_subscriptions(
+    ocm_api: OCMBaseClient,
+    subscription_ids: list[str],
+    cluster_filter: Optional[Filter] = None,
+) -> list[OCMCluster]:
+    if not subscription_ids:
+        return []
+
+    return list(
+        get_clusters_for_subscriptions(
+            ocm_api=ocm_api,
+            subscription_filter=Filter().is_in("id", subscription_ids),
+            cluster_filter=cluster_filter,
+        ).values()
+    )
+
+
+def discover_clusters_for_organizations(
+    ocm_api: OCMBaseClient,
+    organization_ids: list[str],
+    cluster_filter: Optional[Filter] = None,
+) -> list[OCMCluster]:
+    if not organization_ids:
+        return []
+
+    return list(
+        get_clusters_for_subscriptions(
+            ocm_api=ocm_api,
+            subscription_filter=Filter().is_in("organization_id", organization_ids),
+            cluster_filter=cluster_filter,
+        ).values()
+    )
+
+
+def get_clusters_for_subscriptions(
+    ocm_api: OCMBaseClient,
+    subscription_filter: Optional[Filter] = None,
+    cluster_filter: Optional[Filter] = None,
+) -> dict[str, "OCMCluster"]:
+    # get subscription details
+    subscriptions = get_subscriptions(
+        ocm_api=ocm_api,
+        filter=(subscription_filter or Filter()).combine(build_subscription_filter()),
+    )
+    if not subscriptions:
+        return {}
+
+    # get organization labels
+    organization_labels: dict[str, list[OCMOrganizationLabel]] = defaultdict(list)
+    for label in get_organization_labels(
+        ocm_api=ocm_api,
+        filter=Filter().is_in(
+            "organization_id",
+            {s.organization_id for s in subscriptions.values()},
+        ),
+    ):
+        organization_labels[label.organization_id].append(label)
+
+    result: dict[str, OCMCluster] = {}
+    cluster_search_filter = (
+        cluster_ready_for_app_interface()
+        .combine(cluster_filter)
+        .is_in("subscription.id", subscriptions.keys())
+    )
+    chunk_size = 100
+    for filter_chunk in cluster_search_filter.chunk_by(
+        "subscription.id", chunk_size, ignore_missing=True
+    ):
+        for cluster_dict in ocm_api.get_paginated(
+            api_path="/api/clusters_mgmt/v1/clusters",
+            params={"search": filter_chunk.render()},
+            max_page_size=chunk_size,
+        ):
+            subscription = subscriptions[cluster_dict["subscription"]["id"]]
+            result[cluster_dict["subscription"]["id"]] = OCMCluster(
+                **cluster_dict,
+                subscription_id=cluster_dict["subscription"]["id"],
+                organization_id=subscription.organization_id,
+                product_id=cluster_dict["product"]["id"],
+                region_id=cluster_dict["region"]["id"],
+                api_url=cluster_dict["api"]["url"],
+                console_url=cluster_dict["console"]["url"],
```

Comment:
Wondering if it makes sense to wrap these in a try-catch and raise sth like `OcmApiError` and log more understandable error if OCM API returned sth we didnt expect.

Reply:
good point. i reworked this code to rely exclusively on pydantic models to parse and encode data. so any error because of invalid/incomplete data is going to be very explicit because of pydantics way to report validation issues.

do you consider this sufficient?